sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureCatalog.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.catalyst.analysis.NoSuchProcedureException;
+import org.apache.spark.sql.connector.catalog.procedures.UnboundProcedure;
+
+/**
+ * A catalog API for working with procedures.
+ *
+ * @since 4.1.0
+ */
+@Evolving
+public interface ProcedureCatalog extends CatalogPlugin {
+  /**
+   * Load a procedure by {@link Identifier identifier} from the catalog.
+   *
+   * @param ident a procedure identifier
+   * @return the loaded unbound procedure
+   * @throws NoSuchProcedureException if the procedure does not exist
```

Comment:
Do we really need a specific exception type? I think we should give the implement freedom to throw whatever error they want, if they use Spark's error framework. e.g. implementation may throw access denied error.

When Spark invokes this method, we should throw the error as it is if it's `SparkThrowable` with the error class defined. If not, Spark should throw `FAILED_LOAD_PROCUDURE` and set the original error as the cause.

Reply:
According to my experience, enforcing a certain exception type brings more troubles to the implementation, as it has to try-catch errors from the underlying libraries/clients that already use Spark error class, to rethrow the error with the required exception type. It also stops the implementation of customizing their own "NOT FUOND" errors.

We should also relax this exception type requirement in other DS v2 connector APIs later on, to be consistent.

Reply:
This makes sense for me as long as we will be consistent. I'll update this PR.

Reply:
Updated.